peps/pep-0750.rst

```diff
@@ -53,638 +34,500 @@ deferred pending further experience with f-strings. Work on this PEP was
-    assert str(result) = '<header>' # etc.
+        @property
+        def message(self) -> str:
+            # Use the f() function from the previous example
+            return f(self.template)
 
-This ``html`` tag function might have the following signature:
+        @property
+        def values(self) -> Mapping[str, Any]:
+            return {
+                arg.expr: arg.value
+                for arg in self.template.args
+                if isinstance(arg, Interpolation)
+            }
 
-.. code-block:: python
+        @property
+        def data(self) -> Mapping[str, Any]:
+            return {"message": self.message, "values": self.values}
 
-    def html(*args: Decoded | Interpolation) -> HTML:
-        ...
+        def __str__(self) -> str:
+            return json.dumps(self.data)
 
-The ``HTML`` return class might have the following shape as a ``Protocol``:
+    _ = TemplateMessage  # optional, to improve readability
+    foo, bar, num, fnum = "bar", "baz", 123, 123.456
+    logging.info(_(t"The {foo} is {bar} and {num} is close to {fnum:.2f}"))
 
-.. code-block:: python
+    # Outputs (without indentation):
+    # {
+    #    "message": "The bar is baz and 123 is close to 123.46"
+    #    "values": {"foo": "bar", "bar": "baz", "num": 123, "fnum": 123.456}
+    # }
 
-    @runtime_checkable
-    class HTML(Protocol):
-        tag: str
-        attrs: dict[str, Any]
-        children: Sequence[str | HTML]
+Use of template strings gives us a more structured way to define the message
 
-In summary, the returned instance can be used as:
 
-- A string, for serializing to the final output
-- An iterable, for working with WSGI/ASGI for output streamed and evaluated
-  interpolations *in the order* they are written out
-- A DOM (data) structure of nested Python data
+Example: HTML Templating
+-------------------------
 
-In each case, the result can be lazily and recursively composed in a safe fashion, because
-the return value isn't required to be a string. Recommended practice is that
-return values are "passive" objects.
+**TODO**
 
-What benefits might come from returning rich objects instead of strings? A DSL for
-a domain such as HTML templating can provide a toolchain of post-processing, as
-`Babel <https://babeljs.io>`_ does for JavaScript
-`with AST-based transformation plugins <https://babeljs.io/docs/#pluggable>`_.
-Similarly, systems that provide middleware processing can operate on richer,
-standard objects with more capabilities. Tag string results can be tested as
-nested Python objects, rather than string manipulation. Finally, the intermediate
-results can be cached/persisted in useful ways.
 
 Tool Support
 ============
 
-Python Semantics in Tag Strings
--------------------------------
+Python Semantics in Template Strings
+------------------------------------
+
+**TODO** still trying to decide how to revise this section...
 
 Python template languages and other DSLs have semantics quite apart from Python.
```

Comment:
Minor: Add "Currently, " just to emphasize it's the status quo, not the post-PEP world.

Reply:
@pauleveritt What do you think about getting rid of this section entirely? I added a bit of a hopeful nod to future tool support at very end of the "Rejected Ideas" section.

Reply:
Agreed.